### PR TITLE
Fix xDS client for multiple watchers

### LINKED ISSUE #28468: XdsClient: Multiple watchers on the same resource
Noticed through flakiness in the recently introduced XdsRbacTest suite.

https://source.cloud.google.com/results/invocations/c4261ea1-e132-49c2-bf67-dd739abf8c9e/targets/%2F%2Ftest%2Fcpp%2Fend2end%2Fxds:xds_end2end_test@poller%3Depoll1/tests

When there are multiple watchers for the same resource on an XdsClient object, cancellation of one of the watchers leads to unsubscription for the resource. The correct behavior should be to unsubscribe only when there are no more watchers.

### PR DESCRIPTION
**Pull request by yashykt**
Fixes #28468 
The added test fails before the change.

### CODE REVIEW

**Comment by markdroth** [comment]
Does this actually need to depend on #28467?  I would think that we could fix this and add the new test without affecting the other test changes.

Not a big deal either way, though.

**Comment by yashykt** [comment]
This raises the failure rate of the tests fixed by #28468 to be much higher than what it is right now without

**Comment by markdroth** [comment]
> This raises the failure rate of the tests fixed by #28468 to be much higher than what it is right now without

Are you saying that this PR without #28467 causes the tests in #28468 to be worse?  If so, it seems like we could just merge #28467 before this PR.

File: test/cpp/end2end/xds/xds_end2end_test.cc [hunk 1]
```diff
@@ -10291,6 +10291,35 @@ TEST_P(XdsRbacTestWithRouteOverrideAlwaysPresent,
   SendRpc([this]() { return CreateInsecureChannel(); }, {}, {});
 }
 
+using XdsRbacTestWithRouteOverrideAndRdsTestingEnabled = XdsRbacTest;
```
[the remark this answers is not in the document]
Comment (yashykt) [reply test/cpp/end2end/xds/xds_end2end_test.cc]: Done! Either way works imo

**Comment by yashykt** [comment]
> Are you saying that this PR without #28467 causes the tests in #28468 to be worse? If so, it seems like we could just merge #28467 before this PR.

Yup, that's what I plan to do. Doing this just made my local testing simpler.

**Review by markdroth — APPROVED** [review]
Looks great!  I assume you've verified that the test fails without the fix.

**Comment by yashykt** [comment]
> Looks great! I assume you've verified that the test fails without the fix.

Yes it does

**Pull request merged**